include/irr/video/IGPUDescriptorSet.h

```diff
@@ -0,0 +1,22 @@
+#ifndef __IRR_I_GPU_DESCRIPTOR_SET_H_INCLUDED__
+#define __IRR_I_GPU_DESCRIPTOR_SET_H_INCLUDED__
+
+#include "irr/asset/IDescriptorSet.h"
+#include "IGPUBuffer.h"
+#include "ITexture.h"
+
+namespace irr { namespace video
+{
+
+class IGPUDescriptorSet : public asset::IDescriptorSet<IGPUBuffer, ITexture>, public core::IReferenceCounted
```

Comment:
@Crisspl you should not make a full descriptor set (together with contents) via the constructor.

On vulkan you have to `updateDescriptorSet` after creating.

The CPU object should also have mutators and those should be used instead of the constructor to fill the object.

Reply:
yea, but it's useless isn't it? I'd have to use `vector` instead our `smart_refctd_dynamic_array` and/or create some weird `friend`ships between driver and descriptor set classes....

Reply:
> I'd have to use vector instead of our smart_refctd_dynamic_array

Why? you know up-front how many descriptors will be in a DescriptorSet.

> create some weird friendships between driver and descriptor set classes....

Not really, just do
```cpp
type COpenGLDriver::updateDescriptorSet(IGPUDescriptorSet* set, Args... args)
{
   // `updateDescriptorSet` only defined in COpenGLDescriptorSet not in IGPUDescriptorSet
   return static_cast<COpenGLDescriptorSet*>(set)->updateDescriptorSet(args...);
}
```

Reply:
> yes in that case it would be exceptionally OK to keep a pointer to IDriver or COpenGLDriver* in the GPU object implementation.

but for what? for what driver would be used?

Reply:
>Why? you know up-front how many descriptors will be in a DescriptorSet.

yea, i dont understand concept of updating descriptor sets.. So first i allocate empty descriptor set and then fill it with descriptors? And i must update it only once because number of descriptors is known in the moment of allocation? And i cant update it any more? It's completely useless API then.

Reply:
> but for what? for what driver would be used?

Mostly for the Vulkan implementation since you need a `VkDevice` parameter to `vkUpdateDescriptorSets`.

> yea, i dont understand concept of updating descriptor sets.. So first i allocate empty descriptor set and then fill it with descriptors?

You can fill it with descriptors 1 binding at a time (or several) see the struct 
https://www.khronos.org/registry/vulkan/specs/1.1-extensions/man/html/VkWriteDescriptorSet.html
or you can copy from other descriptor sets (or several)
https://www.khronos.org/registry/vulkan/specs/1.1-extensions/man/html/VkCopyDescriptorSet.html

You can update at any binding if you look at the above structs

> And i must update it only once because number of descriptors is known in the moment of allocation?

Nope you can update as many times as you like, see the structs above. Order of update doesn't matter as long as the updates don't overlap/overwrite.

> And i cant update it any more? It's completely useless API then.

The CPU version you can update as much as you like, the GPU version you can update until you bind it before a draw for the first time, or you create it with an extension that allows for "update after bind". But main logic here under Vulkan is that you cannot update the descriptor set while its being used for drawing, before or after is fine.

Reply:
@Crisspl latest commit does this?

Reply:
yea